core/src/Streamly/Internal/Data/Fold/Type.hs

```diff
@@ -12,45 +12,115 @@
+--      (s -> m b)               -- extract
+-- @
 --
--- == Terminating Folds with Leftovers
+-- The fold driver stops driving the fold as soon as the fold returns a @Done@.
+-- @extract@ is required only if the fold has not stopped yet and the input
+-- ends. @extract@ can never be called if the fold is @Done@.
 --
--- The next upgrade after terminating folds is terminating folds with leftover
--- inputs.  Consider the example of @takeWhile@ operation, it needs to inspect
--- an element for termination decision. However, it does not consume the
--- element on which it terminates. To implement @takeWhile@ a terminating fold
--- will have to implement a way to return unconsumed input to the fold driver.
+-- Notice that the @initial@ of `Fold1` type does not return a "Step" type,
+-- therefore, it cannot say "Done" in initial. It always has to consume at
+-- least one element before it can say "Done" for termination, via the @step@
+-- function.
 --
--- Single element leftover case is the most common and its easy to implement it
--- in terminating folds using a @Done1@ constructor in the 'Step' type which
+-- == Folds that terminate after 0 or more input
+--
+-- The `Fold1` type makes combinators like @take 0@ impossible to implement
+-- because they need to terminate even before they can consume any elements at
+-- all.  Implementing this requires the @initial@ function to be able to return
+-- @Done@.
+--
+-- @
+-- data Fold m a b =
+--   forall s. Fold
+--      (s -> a -> m (Step s b)) -- step
+--      (m (Step s b))           -- initial
+--      (s -> m b)               -- extract
+-- @
+--
+-- This is also required if we want to compose terminating folds using an
+-- Applicative or Monadic composition. @pure@ needs to yield an output without
+-- having to consume an input.
+--
+-- @initial@ now has the ability to terminate the fold without consuming any
+-- input based on the state of the monad.
+--
+-- In some cases it does not make sense to use a fold that does not consume any
+-- items at all, and it may even lead to an infinite loop. It might make sense
+-- to use a `Fold1` type for such cases because it guarantees to consume at
+-- least one input, therefore, guarantees progress. For example, in
+-- classifySessionsBy or any other splitting operations it may not make sense
+-- to pass a fold that never consumes an input. However, we do not have a
+-- separate Fold1 type for the sake of simplicity of types/API.
+--
+-- Adding this capability adds a certain amount of complexity in the
+-- implementation of fold combinators. @initial@ has to always handle two cases
+-- now.  We could potentially not implement this in folds to keep fold
+-- implementation simpler, and these use cases can be transferred to the parser
+-- type. However, it would be a bit inconvenient to not have a `take` operation
+-- or to not be able to use `take 0` if we have it. Also, applicative and
+-- monadic composition of folds would not be possible.
```

Comment:
This is more like developer notes.

Reply:
It's fine I guess.

Reply:
These are developer notes. It is doc of internal module, not supposed to be visible in user docs.